### DISCUSSION TITLE
Trying to attach debugger while developing extensions

### DISCUSSION THREAD

**Discussion by keesvanbemmel** [directus/directus#8386] [Archive]
Hey there,

I'm trying to attach my javascript debugger to my directus instance (which is running in a container) using [this walkthrough](https://www.bigbinary.com/blog/debug-nodejs-app-running-in-a-docker-container).

My dev script looks like this: `yarn install; npx directus bootstrap; npx directus schema apply ./schema.yaml --yes; npx nodemon --inspect=0.0.0.0:9229 --legacy-watch --exec directus start`

While running this I'm getting this error: `error: unknown option '--inspect=0.0.0.0:9229'`

I'm not looking specifically for a fix to this error, but I'd really like to debug my extensions while developing. Any tips tricks on this error and/or getting my vscode debugger attached?

Thanks!

**Comment by nickrum** [comment]
I haven't tried this myself, but I think you have to pass the `--inspect` option to the directus node process. You could try something like this: `... npx nodemon --legacy-watch --exec directus start -- --inspect=0.0.0.0:9229`

**Reply by keesvanbemmel** [reply]
Thanks for the tip! Unfortunately, it errors out with the exact same error:

```
hub         | success Already up-to-date.
hub         | 13:30:30 ✨ Loaded extensions: scoring-orderlines
hub         | 13:30:30 ✨ Initializing bootstrap...
hub         | 13:30:30 ✨ Database already initialized, skipping install
hub         | 13:30:30 ✨ Running migrations...
hub         | 13:30:30 ✨ Done
hub         | 13:30:36 ✨ Loaded extensions: scoring-orderlines
postgres_1  | 2021-09-28 13:30:37.723 UTC [36] ERROR:  function postgis_version() does not exist at character 8
postgres_1  | 2021-09-28 13:30:37.723 UTC [36] HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
postgres_1  | 2021-09-28 13:30:37.723 UTC [36] STATEMENT:  select postgis_version()
hub         | 13:30:37 ✨ No changes to apply.
hub         | [nodemon] 2.0.12
hub         | [nodemon] to restart at any time, enter `rs`
hub         | [nodemon] watching path(s): *.*
hub         | [nodemon] watching extensions: js,mjs,json
hub         | [nodemon] starting `directus start --inspect=0.0.0.0:9229`
hub         | 13:30:43 ✨ Loaded extensions: scoring-orderlines
hub         | error: unknown option '--inspect=0.0.0.0:9229'
hub         | [nodemon] app crashed - waiting for file changes before starting...
```

**Reply by nickrum** [reply]
In that case you might have to run the Directus start script directly instead of using the directus cli. We should probably add support for debugging extensions to the upcoming extension dev environment.

Edit: Or you could try using the `NODE_OPTIONS` env var.

**Reply by joselcvarela** [reply]
Since sometimes there is some `spawn` on the code to launch other node processes, I usually do this:
`export NODE_OPTIONS=--inspect && npm run dev`

In your case:
`export NODE_OPTIONS=--inspect && yarn install; npx directus bootstrap; npx directus schema apply ./schema.yaml --yes; npx nodemon --legacy-watch --exec directus start`

**Reply by keesvanbemmel** [reply]
That worked, however I'm getting errors that port 9229 is already in use.

The debugger attaches, but will not stop at breakpoints in hooks or endpoints. I'm guessing port 9229 is in use by one of the Directus processes? Like the app instead of the API?

**Reply by joselcvarela** [reply]
Maybe it's because it is attaching debugger to multiple processes. Try this instead (on new terminal, since the other one is already poluted):
`yarn install; npx directus bootstrap; npx directus schema apply ./schema.yaml --yes; export NODE_OPTIONS=--inspect && npx nodemon --legacy-watch --exec directus start`

**Reply by keesvanbemmel** [reply]
Sorry I should've been clear on that, that's what I figured and did, same result.

**Reply by joselcvarela** [reply]
Huumm I am suspicious about one thing: I think you are attaching debbuger to `nodemon` then the `exec` command is spawning the Directus process and when trying to attach the debugger it is already attached on `nodemon`.
How about this `yarn install; npx directus bootstrap; npx directus schema apply ./schema.yaml --yes; npx nodemon --legacy-watch --exec NODE_OPTIONS=--inspect directus start` ?

**Reply by keesvanbemmel** [reply]
@joselcvarela you are my hero of the day, that worked like a charm. Indeed I'm guessing the same: the inspector was running on the nodemon process, with your command it isn't and now the debugger attaches fine!

If you could post your answer not as a reply but as an actual answer I'll mark it as the answer ffr!

**Comment by joselcvarela** [comment]
@keesvanbemmel no problem we are here to help 😉

Answer:
https://github.com/directus/directus/discussions/8386#discussioncomment-1400491

**Comment by maxbritto** [comment]
Note for people trying to use a debugger for regular directus prod folder copied locally on your machine, here is a simple way, based on the tips given in this discussion : 
## 1 - Add a dev configuration in your `package.json` : 
```
"scripts": {
    "dev": "npx nodemon --legacy-watch --exec NODE_OPTIONS=--inspect directus start",
    "start": "npx directus start",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
```

## 2 - Run your instance using that `dev` configuration
1. Go to your project folder
2. Execute `npm  run dev`

## 3 - Attach your debugger 
In VSCode for example, use the command palette to attach to an existing process (https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_attaching-to-nodejs)
After 2 seconds, you should see *"Debugger attached"* in your directus console

## 4 - Use your debugger
Add breakpoints and load any URL or send requests to your instance : `http://localhost:8055/custom/your-endpoint`.
I've used it successfully with custom endpoints and it should work for other things too.

**Reply by LudovicSterlin** [reply]
>  regular directus prod folder copied locally on your machine

I am not sure what you mean by that... Did you clone the whole [directus repo](https://github.com/directus/directus) and run the app and the api ?

**Reply by maxbritto** [reply]
No, I meant a regular CLI install of directus, as explained in the doc : with the command ˋnpx directus initˋ
But this was my first directus server, now I’m leaning towards using docker compose to create quickly new directus instances.
